stage0/README.md

```diff
@@ -37,3 +64,113 @@ To use the binary, pass it to `qemu -bios`; for example:
+`fw_cfg` entries:
+
+```shell
+qemu-system-x86_64 [...] -fw_cfg name=opt/stage0/elf_kernel,file=/path/to/kernel \
+                         -fw_cfg name=opt/stage0/initramfs,file=/path/to/initramfs \
+                         -fw_cfg name=opt/stage0/cmdline,string=console=ttyS0
+```
+
+Unfortunately we had to implement custom `fw_cfg` entries; the standard
+`-kernel` flag won't work as QEMU may load files into guest memory, but it will
+not ask the PSP to encrypt the memory. If we don't provide a `-kernel` flag,
+QEMU will reject `-initrd` and `-append` flags; thus, we had to reimplement
+those as well.
+
+This approach is supported under SEV and SEV-ES. (Probably also SEV-SNP, but the
+public releases of QEMU do not support SEV-SNP yet.)
+
+## Internals
+
+### Memory layout
+
+Stage0 maps the first 1 GiB of physical memory using identity mapping before
+handing control over to the kernel. The stage0 binary itself is mapped into
+memory just below the end of 4 GiB address space, like any other BIOS ROM.
+
+```text
+￼              |                      ...                       |
+￼0x1_0000_0000 +------------------------------------------------+ 4GiB
+￼              |               Stage 0 ROM Image                |
+ ￼ 0xFFFE_0000 +------------------------------------------------+ 4GiB - 128MiB
+￼              |                    PCI Hole                    |
+￼  0xC000_0000 +------------------------------------------------+ 3GiB
+￼              |                  Unmapped RAM                  |
+￼  0x4000_0000 +------------------------------------------------+ 1GiB
+￼              |                                                |
+                                      ...
+               |                                                |
+￼              +------------------------------------------------+
+￼              |                                                |
+￼              |               Restricted Kernel                |
+￼              |                                                |    Kernel
+  ￼  0x20_1000 +------------------------------------------------+ <- Entry Point
+￼              |          Restricted Kernel ELF Header          |
+  ￼  0x20_0000 +------------------------------------------------+ 2MiB
+￼              |                                                |
+  ￼   0xA_0000 +------------------------------------------------+ 640KiB
+￼              |     Extended BIOS Data Area (ACPI Tables)      |
+      0x8_0000 +------------------------------------------------+ 512KiB
+￼              |             Stage 0 Stack and Heap             |
+￼              | (Includes Secrets and CPUID pages on SEV-SNP)  |
+  ￼     0x1000 +------------------------------------------------+ 4KiB
+￼              |                 Unmapped Page                  |
+￼          0x0 +------------------------------------------------+
+```
+
+## Future work
+
+- Support for Intel TDX
+- Attest the loaded kernel
```

Comment:
Do we have/want to link an issue?

Reply:
I don't think we have separate issues for these, but otherwise the answer would be "yes".